### PR TITLE
4750 issue 123 refactor contacts effect

### PR DESCRIPTION
**Pull request by njogz**
# Description

This PR addresses the following issues:

- [issue 121](https://github.com/medic/angular10-migration/issues/121)
- [issue 123](https://github.com/medic/angular10-migration/issues/123)
- [issue 129](https://github.com/medic/angular10-migration/issues/129)

medic/cht-core#[number]

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by dianabarsan — COMMENTED** [review]
Postponing this review, as more refactors are necessary based on some missing features, which might contour the final code differently.

**Comment by njogz** [comment]
@dianabarsan I have made the refactor. Please have a look when you can. Thanks.

File: webapp/src/ts/effects/contacts.effects.ts [hunk 1]
```diff
@@ -105,6 +108,8 @@ export class ContactsEffects {
       exhaustMap(([, selectedContact, forms]) => {
         return from(this.contactViewModelGeneratorService.loadReports(selectedContact, forms)).pipe(
           map(reports => {
+            const selected = { ...selectedContact, reports };
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:111]: I don't think we should treat effects like reducers. 
We shouldn't "reduce" the selected contact here. 
I think `loadSelectedContactTargetDoc` should get no parameters.

File: webapp/src/ts/effects/contacts.effects.ts [hunk 2]
```diff
@@ -116,50 +121,47 @@ export class ContactsEffects {
...
-          .get(selected)
-          .then((tasks) => {
-            return this.contactsActions.updateSelectedContactsTasks(tasks);
-          });
-        return from(this.contactSummaryService.get(selected.doc, selected.reports, selected.lineage)).pipe(
-          map(summary => {
-            this.contactsActions.setContactsLoadingSummary(false);
-            return this.contactsActions.updateSelectedContactSummary(summary);
+      ofType(ContactActionList.loadSelectedContactTargetDoc),
+      exhaustMap(({ payload: { selected } }) => {
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:127]: This action doesn't need a payload, and can get the selectedContact directly from the state: 
`withLatestFrom(this.store.pipe(select(Selectors.getSelectedContact))`

File: webapp/src/ts/effects/contacts.effects.ts [hunk 3]
```diff
@@ -116,50 +121,47 @@ export class ContactsEffects {
...
+        return from(this.targetAggregateService.getCurrentTargetDoc(selected)).pipe(
+          map(targetDoc => {
+            const selectedContact = { ...selected, targetDoc };
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:130]: Like above, I don't think effects should do the work of reducers.   

What I think would makes things clearer is: 
- call `receiveSelectedContactTargetDoc` action with the targetDoc param and have a reducer save that in the store
- call `loadSelectedContactSummary` which doesn't need any payload
- the effect "updateSelectedContactSummary"  should wait for `loadSelectedContactSummary` and get the `selectedContact` from the store and load the contact summary as usual

File: webapp/src/ts/effects/contacts.effects.ts [hunk 4]
```diff
@@ -116,50 +121,47 @@ export class ContactsEffects {
...
-          }),
-          catchError(error => {
-            console.error('Error loading target doc', error);
-            return of(this.globalActions.unsetSelected());
-          }),
-        );
+      ofType(ContactActionList.loadSelectedContactSummary),
+      exhaustMap(({ payload: { selected } }) => {
+        this.contactsActions.setContactsLoadingSummary(true);
+        this.tasksForContactService
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:148]: I think loading tasks should be done in a different effect, they have nothing to do with contact-summary.

File: webapp/src/ts/effects/contacts.effects.ts [hunk 5]
```diff
@@ -116,50 +121,47 @@ export class ContactsEffects {
...
+          .get(selected)
+          .then((tasks) => {
+            return this.contactsActions.updateSelectedContactsTasks(tasks);
+          });
+        return from(this.contactSummaryService
+          .get(selected.doc, selected.reports, selected.lineage, selected.targetDoc))
```
Comment (dianabarsan) [inline webapp/src/ts/effects/contacts.effects.ts:154]: this indentation is kinda off. it looks like you're chaining from the `get` but you're instead chaining from the `from`.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! I have a few comments about consistency and clarity and differentiation between effects and reducers.

File: webapp/src/ts/effects/contacts.effects.ts (hunk 3 above)
Comment (njogz) [reply webapp/src/ts/effects/contacts.effects.ts:130]: Great suggestion. I have refactored as suggested.

File: webapp/src/ts/effects/contacts.effects.ts (hunk 4 above)
Comment (njogz) [reply webapp/src/ts/effects/contacts.effects.ts:148]: Done.

**Review by dianabarsan — APPROVED** [review]
Nice work! Thanks!

**Comment by dianabarsan** [comment]
The build is failing with an eslint error @njogz 
> /home/runner/work/cht-core/cht-core/webapp/tests/karma/ts/effects/contacts.effects.spec.ts
  340:9  error  'updateSelectedContactsTasks' is assigned a value but never used  @typescript-eslint/no-unused-vars

Can you, please, fix the linting error?

**Pull request merged**